### PR TITLE
Front end for authorisation done with a login page

### PR DESCRIPTION
**Pull request by mateenparkar**
image

### CODE REVIEW

**Comment by TomBillingtonUK** [comment]
Missing unit tests

File: test/unit/authServiceTest.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        mock.onPost('http://localhost:8080/api/login', loginData).reply(401);
+
+        try {
+            await login(loginData);
+          } catch (error:any) {
+            console.log(error)
+            expect(error.message).to.equal('Could not login');
+        }
+    });
+});
```
Comment (TomBillingtonUK) [inline test/unit/authServiceTest.ts:50]: Missing a unit test for the 401 path

File: test/ui/LoginTests.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+        await driver.findElement(By.id('submit')).click();
+
+        await driver.findElement(By.id('invalid-user-login')).getText().then(function(value:any) {
+            chai.assert.equal(value, 'Your email or password combination is incorrect');
+        });
+
+        await driver.quit();
+    });
+});
```
[the remark this answers is not in the document]
Comment (mateenparkar) [reply test/ui/LoginTests.ts:45]: For a ui test aren't we just testing the error that displays on the front end? Which is the "Your email or password combination is incorrect" error so we wouldn't have to test the Could not login error?

File: src/controller/authController.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import type { Application, Request, Response } from "express"
+import { ActiveSession, Credentials } from "../model/auth"
+
+import { login } from "../service/authService"
+
+export const authController = (app:Application) =>{
```
Comment (shaunganley) [inline src/controller/authController.ts:6]: it looks like this file isn't used

File: src/service/authService.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { ActiveSession, Credentials } from "../model/auth";
+
+import axios, { AxiosError } from 'axios'
+
+export const login = async function(credentials: Credentials): Promise<ActiveSession> {
+    try {
+        const response = await axios.post('http://localhost:8080/api/login/', credentials);
```
Comment (shaunganley) [inline src/service/authService.ts:7]: still needs to be addressed

File: test/ui/LoginTests.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+describe('Login Test', async () => {
+    it('Login with a valid user', async () => {
+        const driver = new Builder().
+            withCapabilities(Capabilities.chrome()).
+            build();
+        
+        await driver.get(process.env.UI_TEST_URL as string);
+
+        await driver.findElement(By.id('username')).sendKeys('email@email.com');
+        await driver.findElement(By.id('password')).sendKeys('password');
```
Comment (shaunganley) [inline test/ui/LoginTests.ts:15]: you should pass valid username and password and ENV VARs

File: test/ui/LoginTests.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+        const usernameInput = await driver.findElement(By.id('username'));
+        const enteredValue = await usernameInput.getAttribute('value');
+
+        chai.assert.equal(enteredValue, 'email@email.com');
+
+
+        await driver.findElement(By.id('submit')).click();
```
Comment (shaunganley) [inline test/ui/LoginTests.ts:23]: this should check that you get redirected after successful login

File: test/unit/authServiceTest.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    it('should return login from response', async () => {
+        const mock = new MockAdapter(axios);
+        const loginData:Credentials = {
+            username: "johndoe@gmail.com",
+            password: "password"
+        }
+
+        const responseData = "token"
+
+        mock.onPost('http://localhost:8080/api/login/', loginData).reply(200, responseData);
```
Comment (shaunganley) [inline test/unit/authServiceTest.ts:24]: don't hardcode URL

File: test/unit/authServiceTest.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    });
+
+    it('should throw an error for failed login', async () => {
+        const mock = new MockAdapter(axios);
+        const loginData = {
+            username: "johndoe@gmail.com",
+            password: "password"
+        };
+    
+        mock.onPost('http://localhost:8080/api/login/', loginData).reply(401, {}); // Simulate a 401 error
```
Comment (shaunganley) [inline test/unit/authServiceTest.ts:52]:
```suggestion
        mock.onPost('http://localhost:8080/api/login/', loginData).reply(500);
```

**Review by shaunganley — CHANGES_REQUESTED** [review]

**Review by shaunganley — APPROVED** [review]

**Review by Adaray786 — APPROVED** [review]

**Pull request merged**